lua/cmp/view/statusline_entries_view.lua

```diff
@@ -0,0 +1,296 @@
+          })
+
+          if i == self.selected_index then
+            vim.api.nvim_buf_set_extmark(buf, statusline_entries_view.ns, 0, location, {
+              end_line = 0,
+              end_col = location + view['abbr'].bytes,
+              hl_group = 'PmenuSel',
+              hl_mode = 'combine',
+              ephemeral = true,
+            })
+          end
+
+          for _, m in ipairs(e.matches or {}) do
+            vim.api.nvim_buf_set_extmark(buf, statusline_entries_view.ns, 0, location + m.word_match_start - 1, {
+              end_line = 0,
+              end_col = location + m.word_match_end,
+              hl_group = m.fuzzy and 'CmpItemAbbrMatchFuzzy' or 'CmpItemAbbrMatch',
+              hl_mode = 'combine',
+              ephemeral = true,
+            })
+          end
+
+          location = location + view['abbr'].bytes + 1
+        end
+      end
+    end,
+  })
+
+  autocmd.subscribe(
+    'CompleteChanged',
+    vim.schedule_wrap(function()
+      if self:visible() and vim.fn.pumvisible() == 1 then
+        self:close()
+      end
+    end)
+  )
+  return self
+end
+
+statusline_entries_view.close = function(self)
+  self.entries_win:close()
+  self:init()
+end
+
+statusline_entries_view.ready = function()
+  return vim.fn.pumvisible() == 0
+end
+
+statusline_entries_view.on_change = function(self)
+  self.active = false
+end
+
+statusline_entries_view.open = function(self, offset, entries)
+  self.offset = offset
+  self.entries = {}
+  self.last_displayed_indices = {}
+
+  -- Apply window options (that might be changed) on the custom completion menu.
+  self.entries_win:option('winblend', vim.o.pumblend)
+
+  -- local entries_buf = self.entries_win:get_buffer()
+  local dedup = {}
+  local preselect = 0
+  local i = 1
+  for _, e in ipairs(entries) do
+    local view = e:get_view(offset, 0)
+    if view.dup == 1 or not dedup[e.completion_item.label] then
+      dedup[e.completion_item.label] = true
+      table.insert(self.entries, e)
+      if preselect == 0 and e.completion_item.preselect then
+        preselect = i
+      end
+      i = i + 1
+    end
+  end
+
+  self.entries_win:open({
+    relative = 'editor',
+    style = 'minimal',
+    row = vim.api.nvim_win_get_height(0),
+    col = 0,
+    width = vim.api.nvim_win_get_width(0),
+    height = 1,
+    zindex = 1001,
+  })
+  if preselect > 0 and config.get().preselect == types.cmp.PreselectMode.Item then
+    self:_select(preselect, { behavior = types.cmp.SelectBehavior.Select })
+  elseif not string.match(config.get().completion.completeopt, 'noselect') then
+    self:_select(1, { behavior = types.cmp.SelectBehavior.Select })
+  else
+    self:_select(nil, { behavior = types.cmp.SelectBehavior.Select })
+  end
+end
+
+statusline_entries_view.abort = function(self)
+  feedkeys.call('', 'n', function()
+    self:close()
+  end)
+end
+
+statusline_entries_view.draw = function(self)
+  local entries_buf = self.entries_win:get_buffer()
```

Comment:
Probably, I'll try to refactor it using the `&sidescrolloff` and `vim.api.nvim_win_set_cursor` for managing viewport after merge this PR.